### ISSUE TITLE
qBittorrent v2 API unstable

### ISSUE THREAD

**Issue by kevin-stuart** [erickok/transdroid#567]
Apologies if this could be considered a duplicate: I replied to the closed issue [#534](https://github.com/erickok/transdroid/issues/534) about 6 weeks ago, but it has not gotten any responses, unlike some other newly opened issues.  The original issue also lacked a bit of  direction.

Basically, superficially it appears that the qBittorrent v2 API is working, but I think if it is used more extensively it is not true at all. I think the problem might stem from either Transdroid trying to detect the API version and there could be some kind of race condition.

From my orginal report: If you try to quickly access qBittorrent with Transdroid, you get 404 with a call to the v1 login and this seems to ruin valid results that Transdroid gets with the v2 API. Here is a log from my initial report:

```
x.x.x.x - - [22/Jul/2020:15:07:08 +0200] "POST /api/v2/auth/login HTTP/1.1" 200 3 "-" "Transdroid Torrent Connect"
x.x.x.x - - [22/Jul/2020:15:07:09 +0200] "POST /login HTTP/1.1" 404 0 "-" "Transdroid Torrent Connect"
x.x.x.x - - [22/Jul/2020:15:07:09 +0200] "POST /api/v2/auth/login HTTP/1.1" 200 3 "-" "Transdroid Torrent Connect"
x.x.x.x - - [22/Jul/2020:15:07:09 +0200] "POST /api/v2/app/version HTTP/1.1" 200 6 "-" "Transdroid Torrent Connect"
x.x.x.x - - [22/Jul/2020:15:07:09 +0200] "POST /api/v2/sync/maindata?rid=0 HTTP/1.1" 200 2151 "-" "Transdroid Torrent Connect"
```

As you can see there is a call to `/login` instead of `/api/v2/auth/login`. Even though there is a successful port in the last line of the log, Transdroid will display no content and the red error bar is displayed.

It is hard to reproduce reliably, but basically, if you make enough fast calls it happens very regularly. Anyway, I consider this to be very problematic and want to update qBittorrent soon to use the new API, but Transdroid works very badly with the new API. It would be great if you could acknowledge and investigate the problem. A maintenance release that fixes this problem would be very welcome, maybe a simple fix could be to offer a selection of "qBittorrent API v1" and "qBittorrent API v2+" as client type if it is too hard to track down the exact root cause, since race conditions are hard to track down.

**Comment by erickok** [comment]
What version of Transdroid and qBittorrent are you on? Since the most recent Transdroid release the very first call on a clean startup is to log in to the v2 API (as opposed to before where it tried to get the version number, causing 404s when on a v2 qBit release). However, it seems that after that api/v2/auth/login call (which was http 200 ok) it didn't think the auth was successful, as it tried to log in using the v1 endpoint right after. The 'am I logged in' check is done by checking the returned cookies. Ref https://github.com/erickok/transdroid/blob/master/app/src/main/java/org/transdroid/daemon/adapters/qBittorrent/QBittorrentAdapter.java#L190

Can you check the server if this is on the right time and timezone? Sometimes this causes the cookies to be directly invalid upon first login. Otherwise, is there some way to see the response headers on that first /api/v2/auth/login call? It's weird because seemingly upon the 2nd try the auth was done correctly and everything continues without issues.

**Comment by kevin-stuart** [comment]
As I wrote in the original report, I used qBittorrent 4.2.5 and Transdroid 2.5.19. I think both are still the most recent versions. The time/timezone on the server with the logs is UTC+2 (=CEST) which is correct, this is the same timezone as on the phone running Transdroid.

As for seeing response headers, do you have some kind of verbose logging that could be used for this? I.e. it makes more sense to me to log response headers in the client (like the browser, or here Transdroid) to see what Transdroid gets instead of trying to log the headers in the webserver log, which can be a lot of data and may it may be sensitive.

**Comment by erickok** [comment]
All the logging done by Transdroid you can view via the Help menu -> Error report. I currently do not have an extra verbose log mode; that would be something I'd have to add. I would, alternatively, be happy to debug directly against your instance, but I'd need access. Shoot me a mail at transdroid@2312.nl if you're up to that. Just thinking out loud on how to solve this, as I cannot reproduce this locally yet.

**Comment by kevin-stuart** [comment]
Since 6 weeks have passed, I had to investigate my setup a bit again. Since I only tested it and did not yet migrate to the newest qBittorrent version since Transdroid was unreliable for me, I only had a test instance, which I now upgraded from Ubuntu 18.04 to 20.04.

It felt harder for me to produce the issue and I think it may only occur under special cirumstances; I think if I was on the new API and stayed there, no amount of refreshes caused the problem (which was not the case from memory), but this may be incorrect. OTOH, since I still have my main instance with the old version of qBittorrent that uses the old API, I would get the error if I changed from the instance with the old version to the instance of the new version. Is it possible that there is some kind of renmant when switching servers that some calls will me made with the old v1 API to the enw v2 API in this case?

**Comment by erickok** [comment]
Right, yes... I think it is pretty much incompatible with swithcinug between servers that uses v1, v2 or even legacy APIs as there is internal 'static' state. So yes, most certianly that would cause issues with the conenction, as Transdroid would, for example, think you are using v1 API but then switch config to a v2 API and the calls will fail. 

That is not great. Transdroid should 'reset' the state when switching servers.

**Comment by kevin-stuart** [comment]
OK, so I think we found the root cause of the problem. The switch seems to be the problem. I did not notice it intially when I reported it, but now it seems plausible since I had my old and the new server side by side and just tried it out and then saw the error messages.

**Comment by bwitt** [comment]
I setup a new qBittorrent 4.3.1 server and created a new server in transdroid 2.5.18 to connect to this and am seeing this same issue (nginx logs from proxying)  server timezone is UTC:

```
x.x.x.x - - [04/Dec/2020:07:58:11 +0000] "POST /api/v2/auth/login HTTP/1.1" 200 3 "-" "Transdroid Torrent Connect"
x.x.x.x - - [04/Dec/2020:07:58:12 +0000] "POST /login HTTP/1.1" 404 9 "-" "Transdroid Torrent Connect"
x.x.x.x - - [04/Dec/2020:07:58:12 +0000] "POST /api/v2/auth/login HTTP/1.1" 200 3 "-" "Transdroid Torrent Connect"
x.x.x.x - - [04/Dec/2020:07:58:12 +0000] "POST /api/v2/app/version HTTP/1.1" 200 6 "-" "Transdroid Torrent Connect"
x.x.x.x - - [04/Dec/2020:07:58:12 +0000] "POST /api/v2/sync/maindata?rid=0 HTTP/1.1" 200 15459 "-" "Transdroid Torrent Connect"
```

this happens not every time I interact with that server via transdroid, but maybe it happens every time transdroid does a login (attempting to verify).

the error log in transdroid says (UTC-8):
```
5727 -- Thu Dec 03 23:58:11 PST 2020 -- 3 -- AuthenticationFailure exception: Server rejected our login
```

maybe I can try capturing the response headers from that first `/api/v2/auth/login` call

**Comment by bwitt** [comment]
so this doesn't happen with every login attempt; sometimes / usually everything works fine.

here's the headers from the response to the first `POST /api/v2/auth/login` which is followed by `POST /login`:
```
connection: keep-alive
content-length: 3
content-security-policy: default-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; script-src 'self' 'unsafe-inline'; object-src 'none'; form-action 'self'; frame-ancestors 'self';
content-type: text/plain; charset=UTF-8
date: Fri, 04 Dec 2020 19:49:21 GMT
referrer-policy: same-origin
set-cookie: SID=xxxxxxxxxxxxxxxxxxxxxx; HttpOnly; path=/; SameSite=Strict
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
```

It looks very similar to the headers from the response to the second `POST /api/v2/auth/login`:
```
connection: keep-alive
content-length: 3
content-security-policy: default-src 'self'; style-src 'self' 'unsafe-inline'; img-src 'self' data:; script-src 'self' 'unsafe-inline'; object-src 'none'; form-action 'self'; frame-ancestors 'self';
content-type: text/plain; charset=UTF-8
date: Fri, 04 Dec 2020 19:49:23 GMT
referrer-policy: same-origin
set-cookie: SID=yyyyyyyyyyyyyyyyyyyyyyy; HttpOnly; path=/; SameSite=Strict
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
```

they seem identical except for the date and the SID cookie.

**Comment by kevin-stuart** [comment]
So... a few month ago I concluded that switching may be the problem. It is not. There seems to be some kind of other bug as confirmed by bwitt.

The most reliable way for me to reproduce the bug is the following: Open Transdroid and use the Android task switcher to swipe up Transdroid to hopefully kill the app. Open it again immediately and wait for first load to finish. Most of the time it will work, but *not always*. As soon as the load animation finishes, hit refresh. It either seems to take a lot longer than usual or fail very quickly. I think it *does not* fail once the app is open and has not failed before. Only if you swipe it dead. By using the right timing as I described, I can get this bug very often consistently.

It still think that this looks like some kind of race.
